### PR TITLE
Optimize `isassigned` and inline `jl_array_isassigned`

### PR DESCRIPTION
**Pull request by yuyichao**
Yet another local stale branch....

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

### CODE REVIEW

**Comment by yuyichao** [comment]
Is the nanosoldier running?

**Comment by jrevels** [comment]
Rekicked the server:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: src/ccall.cpp [hunk 1]
```diff
@@ -1741,6 +1741,35 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
...
+        if (jl_is_array_type(aryty)) {
+            jl_value_t *ety = jl_tparam0(aryty);
+            if (jl_isbits(ety)) {
+                emit_expr(aryex, ctx);
+                emit_expr(idxex, ctx);
+                JL_GC_POP();
+                return mark_or_box_ccall_result(ConstantInt::get(T_int32, 1),
+                                                false, rt, unionall, static_rt, ctx);
+            }
+            else if (!jl_has_free_typevars(ety)) { // TODO: jn/foreigncall branch has a better predicate
```
Comment (tkelman) [inline src/ccall.cpp]: didn't that branch get merged already?

Comment (yuyichao) [inline src/ccall.cpp]: This is copied from the comment in `codegen.cpp` for `arrayref` and `arrayset`.

Comment (tkelman) [inline src/ccall.cpp]: is it a stale comment there too then?

Comment (yuyichao) [inline src/ccall.cpp]: No idea and not related to this PR.

Comment (tkelman) [inline src/ccall.cpp]: you're copy pasting code that includes the comment, so yes it is related

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20890#issuecomment-284168875) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b95e1c095791524f7a222bd609b349d3c09f16d1/7df4e49_vs_145eae8/report.md). cc @jrevels

File: src/ccall.cpp (hunk 1 above)
Comment (tkelman) [inline src/ccall.cpp]: bump, what is the better predicate?

**Pull request merged**